pkg/gadgets/socket-collector/tracer/tracer.go

```diff
@@ -0,0 +1,268 @@
+		return ""
+	}
+
+	switch uint8(statusBytes[0]) {
+	case TCP_ESTABLISHED:
+		return "ESTABLISHED"
+	case TCP_SYN_SENT:
+		return "SYN_SENT"
+	case TCP_SYN_RECV:
+		return "SYN_RECV"
+	case TCP_FIN_WAIT1:
+		return "FIN_WAIT1"
+	case TCP_FIN_WAIT2:
+		return "FIN_WAIT2"
+	case TCP_TIME_WAIT:
+		return "TIME_WAIT"
+	case TCP_CLOSE:
+		return "CLOSE"
+	case TCP_CLOSE_WAIT:
+		return "CLOSE_WAIT"
+	case TCP_LAST_ACK:
+		return "LAST_ACK"
+	case TCP_LISTEN:
+		return "LISTEN"
+	case TCP_CLOSING:
+		return "CLOSING"
+	case TCP_NEW_SYN_RECV:
+		return "NEW_SYN_RECV"
+	default:
+		prifixedLog(fmt.Sprintf("Invalid TCP status: %s (%v)", statusHex, statusBytes))
+	}
+
+	return "UNKNOWN"
+}
+
+func parseUDPStatus(statusHex string) string {
+	statusBytes, err := hex.DecodeString(statusHex)
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to decode hex-string UDP status %q: %s", statusHex, err))
+		return ""
+	}
+
+	switch uint8(statusBytes[0]) {
+	case TCP_ESTABLISHED:
+		return "ACTIVE"
+	case TCP_CLOSE:
+		return "INACTIVE"
+	default:
+		prifixedLog(fmt.Sprintf("Unexpected UDP status: %s (%v)", statusHex, statusBytes))
+	}
+
+	return "UNKNOWN"
+}
+
+func getTCPIter() *link.Iter {
+	objs := IterTCPv4Objects{}
+	if err := LoadIterTCPv4Objects(&objs, nil); err != nil {
+		prifixedLog(fmt.Sprintf("Failed to load TCP BPF objects: %v", err))
+		return nil
+	}
+	defer objs.Close()
+
+	it, err := link.AttachIter(link.IterOptions{
+		Program: objs.DumpTcp4,
+	})
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to attach TCP BPF iterator: %v", err))
+		return nil
+	}
+
+	return it
+}
+
+func getUDPIter() *link.Iter {
+	objs := IterUDPv4Objects{}
+	if err := LoadIterUDPv4Objects(&objs, nil); err != nil {
+		prifixedLog(fmt.Sprintf("Failed to load UDP BPF objects: %v", err))
+		return nil
+	}
+	defer objs.Close()
+
+	it, err := link.AttachIter(link.IterOptions{
+		Program: objs.DumpUdp4,
+	})
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to attach UDP BPF iterator: %v", err))
+		return nil
+	}
+
+	return it
+}
+
+func RunCollector(pid uint32, filter *gadgetv1alpha1.ContainerFilter) (string, error) {
+	if err := increaseRlimit(); err != nil {
+		return "", fmt.Errorf("failed to increase memlock limit: %s", err)
+	}
+
+	// TODO: Use filters to collect only TCP or only UDP information
+	iters := []*link.Iter{}
+	iters = append(iters, getTCPIter())
+	iters = append(iters, getUDPIter())
+
+	// TCP/UDP BPF Iters read information from the current namespace thus
+	// we need to change network namespace before opering the iterator
+	netnsHandle, err := netns.GetFromPid(int(pid))
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve network namespace from PID %d: %w", pid, err)
+	}
+	defer netnsHandle.Close()
+
+	if err := netns.Set(netnsHandle); err != nil {
+		return "", fmt.Errorf("failed to set network namespace of PID %d (NetNsHandle %q): %w", pid, netnsHandle.String(), err)
+	}
+
+	var data []Socket
+
+	for _, it := range iters {
+		if it == nil {
```

Comment:
When would the iter be nil? Maybe add this as a comment?

Reply:
Previously yes, it could happen but, as we discussed, I changed functions `getTCPIter` and `getUDPIter` to return error and not `it` cannot be nil within the `for` loop.